### ISSUE TITLE
UrlField.max_length is too short

### ISSUE THREAD

**Issue by balazs-endresz** [wagtail/wagtail#1985]
Wagtail is using the `UrlField` provided by Django, which has `max_length=200`. Sometimes this is not enough, especially for certain embeds.

Could we use a value larger than that for maybe all `UrlField`s in Wagtail? Or what would be the best way to override this on a specific website?

Related to this, someone managed to submit a page with a long url without any validation errors, which then failed when publishing it. But the draft did get saved (along with the long url) but the page editor form didn't render at all (because of an error when rendering the url field and `TEMPLATE_DEBUG=False`). Changing the embed url field length directly in the database fixed this. Although this was an older version of Wagtail, so I'll give it a try on 1.2 with richtext embeds and streamfield embed blocks too.

**Comment by davecranwell** [comment]
If possible I don't think there should be any limit. It wouldn't be unreasonable for implementers to use a UrlField in a model expecting a google maps link. An event, say, where the editor should provide a link to the venue location map. You'd want it to be validated as a URL, but google maps links are in the hundreds of characters, potentially over 255 even.

**Comment by balazs-endresz** [comment]
I thought mysql can handle max_length > 255 but apparently not if it has a unique constraint:

> Any fields that are stored with VARCHAR column types have their max_length restricted to 255 characters if you are using unique=True for the field.

https://docs.djangoproject.com/en/1.9/ref/databases/#character-fields

I guess that makes subclassing `TextField` more appealing.

**Comment by ghostwords** [comment]
Note that this is likely to go unnoticed in development since the default SQLite database backend ignores VARCHAR length limits:

> [(9) What is the maximum size of a VARCHAR in SQLite?](http://www.sqlite.org/faq.html#q9)
> 
> SQLite does not enforce the length of a VARCHAR. You can declare a VARCHAR(10) and SQLite will be happy to store a 500-million character string there. And it will keep all 500-million characters intact. Your content is never truncated. SQLite understands the column type of "VARCHAR(N)" to be the same as "TEXT", regardless of the value of N.

**Comment by thibaudcolas** [comment]
See also #4292:

> Embedding a URL longer than 200 chars results in: `DataError: value too long for type character varying(200)`.

### Steps to Reproduce

1. In a Wagtail project using `StreamField`
2. Add `('embed', EmbedBlock()),` to it
3. Create/edit a page and add an Embed block
4. Add an URL with more than 200 chars (for example: `https://youtu.be/YBpdL9hSac4?module=Widgetize&action=iframe&containerId=VisitOverviewWithGraph&widget=1&moduleToWidgetize=CoreHome&actionToWidgetize=renderWidgetContainer&idSite=1&period=day&date=today&disableLink=1&widget=1`

> From [wagtail/embeds/models.py#L23](https://github.com/wagtail/wagtail/blob/master/wagtail/embeds/models.py#L23) (`url = models.URLField()`) `url` is being created with a default `max_length` of 200. We should be able to safely increase this limit to `max_length=255`. I may still run into issues with Matomo URLs but I can't think of a reason to not allow 255 chars. More would give me more breathing room but that can potentially cause issues with some (older) backends, MySQL < 5.0.3 comes to mind.

**Comment by ababic** [comment]
> In my opinion there shouldn't be any length restriction on embed URLs. I don't see the value in having that restriction to start with, and in many cases shortening the URL to make it comply simply isn't an option.

Conceptually, I agree there's no good reason why the field should be length limited. However, there are some technical limitations that come along with supporting `mysql` as a database backend. Not only is there a maximum supported length for a `VARCHAR` columns to consider, but more restrictive is the fact that the `url` field is used in a `unique_together` constraint on the model, and index keys in `mysql` (needed to enforce this constraint) have a length limit of 1000 bytes. [From the manual](https://dev.mysql.com/doc/refman/8.0/en/myisam-storage-engine.html):

> The maximum key length is 1000 bytes. This can also be changed by changing the source and recompiling. For the case of a key longer than 250 bytes, a larger key block size than the default of 1024 bytes is used.

The simplest fix here is to set `max_length=255` in the `url` field definition to increase the limit a little. 

If we wanted to go higher than that, I think we'd need to rethink how we're enforcing uniqueness on that model. One option might be to add a `hash` field to the `Embed` model (where we'd store an MD5 hash of the combined `url` and `max_width` values) and make that unique instead. This would free us from the technical limitations, but would add significant complexity to something that is currently very simple.

Anyone have any other ideas?

**Comment by dmarcelino** [comment]
> The simplest fix here is to set `max_length=255` in the `url` field definition to increase the limit a little.

I agree with @ababic. While I don't consider this a final fix it would mitigate the issue for some of us.

**Comment by maziar-dandc** [comment]
Instagram embed doesn't work either, due to thumbnail_urls that are too long such as:

https://instagram.fauh1-1.fna.fbcdn.net/v/t51.2885-15/sh0.08/e35/s640x640/81212482_896587687411263_2564978143874747214_n.jpg?_nc_ht=instagram.fauh1-1.fna.fbcdn.net&_nc_cat=103&_nc_ohc=mr0u0N1_U28AX-b_92X&oh=bedb12c1cd591f8570e0833c33d0c24e&oe=5EA87F75

Seeing as this problem is caused by this bug, instead of creating a new issue, I'm just reporting it here, hope that's alright.

**Comment by FlipperPA** [comment]
If we stick with a `CharField` instead of a `TextField`, this can get tricky. MySQL's safe limit is actually 191 characters when you consider UTF-8. See: https://code.djangoproject.com/ticket/18392#comment:3

If we're modifying to a TextField, can we allow relative URLs in a URLField as well? It would be really nice to be able to enter `/my-category/my-blog-post/` so that the URLs work across `http://localhost:8000`, `https://my-dev-server.example.com` and `https://www.example.com`.

**Comment by pySilver** [comment]
Its a DB level issue https://serversforhackers.com/c/mysql-utf8-and-indexing
I remember I've fixed it for one of our projects with specific innodb settings. But here I'm using Postgre, so it shouldn't be an issue.

Any idea how to increase the length of this field? Monkeypatching maybe?

**Comment by pySilver** [comment]
Quick Idea - do not index this field at all, index and md5 hash instead so you'll have your unique constraint and whatever length required urls.

**Comment by GTpyro** [comment]
Here's my monkey patch. To implement, do the following:
1. add the migration file (0005a_patch_wagtailembeds_to_increase_url_length.py) to your wagtail installation. For Heroku, this is `/app/.heroku/python/lib/python3.7/site-packages/wagtail/embeds/migrations/`
2. replace the wagtail embeds models.py. For Heroku, this file is located in 
`/app/.heroku/python/lib/python3.7/site-packages/wagtail/embeds/models.py`
3. run migrations via `python manage.py migrate`

**0005a_patch_wagtailembeds_to_increase_url_length.py**
```python
# Generated by Django 3.0.10 on 2020-10-16 04:44

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ("blog", "0007_auto_20200226_0318"),
        ("wagtailembeds", "0005_specify_thumbnail_url_max_length"),
    ]

    operations = [
        migrations.AlterField(
            model_name="embed",
            name="thumbnail_url",
            field=models.URLField(blank=True, max_length=2000, null=True),
        ),
        migrations.AlterField(
            model_name="embed",
            name="url",
            field=models.URLField(max_length=2000),
        ),
    ]
```

**models.py**
```python
from django.conf import settings
from django.db import models
from django.utils.translation import gettext_lazy as _

EMBED_TYPES = (
    ("video", "Video"),
    ("photo", "Photo"),
    ("link", "Link"),
    ("rich", "Rich"),
)


class Embed(models.Model):
    """
    When embed code is fetched from a provider (eg, youtube) we cache that code
    in the database so we don't need to ask for it again.

    This model is used for caching the embed html code. It also stores some
    metadata which gets displayed in the editor.

    If an instance of this model is deleted, it will be automatically refetched
    next time the embed code is needed.
    """

    url = models.URLField(max_length=2000)
    max_width = models.SmallIntegerField(null=True, blank=True)
    type = models.CharField(max_length=10, choices=EMBED_TYPES)
    html = models.TextField(blank=True)
    title = models.TextField(blank=True)
    author_name = models.TextField(blank=True)
    provider_name = models.TextField(blank=True)
    thumbnail_url = models.URLField(max_length=2000, null=True, blank=True)
    width = models.IntegerField(null=True, blank=True)
    height = models.IntegerField(null=True, blank=True)
    last_updated = models.DateTimeField(auto_now=True)

    class Meta:
        unique_together = ("url", "max_width")
        verbose_name = _("embed")
        verbose_name_plural = _("embeds")

    @property
    def ratio(self):
        if self.width and self.height:
            return self.height / self.width

    @property
    def ratio_css(self):
        ratio = self.ratio
        if ratio:
            return str(ratio * 100) + "%"

    @property
    def is_responsive(self):
        if not getattr(settings, "WAGTAILEMBEDS_RESPONSIVE_HTML", False):
            return False
        return self.ratio is not None

    def __str__(self):
        return self.url
```from django.conf import settings
from django.db import models
from django.utils.translation import gettext_lazy as _

EMBED_TYPES = (
    ("video", "Video"),
    ("photo", "Photo"),
    ("link", "Link"),
    ("rich", "Rich"),
)


class Embed(models.Model):
    """
    When embed code is fetched from a provider (eg, youtube) we cache that code
    in the database so we don't need to ask for it again.

    This model is used for caching the embed html code. It also stores some
    metadata which gets displayed in the editor.

    If an instance of this model is deleted, it will be automatically refetched
    next time the embed code is needed.
    """

    url = models.URLField(max_length=2000)
    max_width = models.SmallIntegerField(null=True, blank=True)
    type = models.CharField(max_length=10, choices=EMBED_TYPES)
    html = models.TextField(blank=True)
    title = models.TextField(blank=True)
    author_name = models.TextField(blank=True)
    provider_name = models.TextField(blank=True)
    thumbnail_url = models.URLField(max_length=2000, null=True, blank=True)
    width = models.IntegerField(null=True, blank=True)
    height = models.IntegerField(null=True, blank=True)
    last_updated = models.DateTimeField(auto_now=True)

    class Meta:
        unique_together = ("url", "max_width")
        verbose_name = _("embed")
        verbose_name_plural = _("embeds")

    @property
    def ratio(self):
        if self.width and self.height:
            return self.height / self.width

    @property
    def ratio_css(self):
        ratio = self.ratio
        if ratio:
            return str(ratio * 100) + "%"

    @property
    def is_responsive(self):
        if not getattr(settings, "WAGTAILEMBEDS_RESPONSIVE_HTML", False):
            return False
        return self.ratio is not None

    def __str__(self):
        return self.url
```

**Comment by pySilver** [comment]
@GTpyro its not really a monkeypatch :) monkeypatch is applied in runtime, without physical file modification.

@ababic what you think about my idea of creating non-editable hash field that would solve unique_together limits?

**Comment by GTpyro** [comment]
Good to know. Thanks for the clarification!

--
Dmitriy

**Comment by ababic** [comment]
@pySilver yep, that's sort of what I had in mind a little bit further up: https://github.com/wagtail/wagtail/issues/1985#issuecomment-469068545

**Comment by pySilver** [comment]
@ababic yeah, I should have missed this message somehow. This is the way I've solved similar issue in our company projects and imho there is no better way to do it.

**Comment by gasman** [comment]
Fixed in #6635
